From 2e94d0e99dcaa75e4debac274230ef92fb56c4d1 Mon Sep 17 00:00:00 2001
From: Laisheng Hu <laisheng@codeaurora.org>
Date: Thu, 12 Dec 2019 12:23:47 +0800
Subject: [PATCH 34/42] omxvideodec:set component to OMX_StateIdle instead of
 OMX_StatePause for flushing

fix CR 2579377: swdec can not seek
this patch only affect our sw dec, won't affect hw or 3rd party dec

Change-Id: I699872d32243092c8d2bce8186677eda22e9536b
Signed-off-by: Laisheng Hu <laisheng@codeaurora.org>
---
 omx/gstomxvideodec.c | 26 +++++++++++++++++++++++++-
 1 file changed, 25 insertions(+), 1 deletion(-)

diff --git a/omx/gstomxvideodec.c b/omx/gstomxvideodec.c
index 8dda8c0..87f24e8 100644
--- a/omx/gstomxvideodec.c
+++ b/omx/gstomxvideodec.c
@@ -2963,6 +2963,8 @@ gst_omx_video_dec_flush (GstVideoDecoder * decoder)
 {
   GstOMXVideoDec *self = GST_OMX_VIDEO_DEC (decoder);
   OMX_ERRORTYPE err = OMX_ErrorNone;
+  GstOMXVideoDecClass *klass = GST_OMX_VIDEO_DEC_GET_CLASS (self);
+  const gchar * cmp_name = klass->cdata.component_name;
 
   GST_DEBUG_OBJECT (self, "Flushing decoder");
 
@@ -2971,7 +2973,29 @@ gst_omx_video_dec_flush (GstVideoDecoder * decoder)
 
   /* 0) Pause the components */
   if (gst_omx_component_get_state (self->dec, 0) == OMX_StateExecuting) {
-    gst_omx_component_set_state (self->dec, OMX_StatePause);
+    if (!strncmp(cmp_name,
+                      "OMX.qti.video.decoder.h263sw",
+                      OMX_MAX_STRINGNAME_SIZE) ||
+      ! strncmp(cmp_name,
+                      "OMX.qti.video.decoder.mpeg4sw",
+                      OMX_MAX_STRINGNAME_SIZE) ||
+      ! strncmp(cmp_name,
+                      "OMX.qti.video.decoder.divxsw",
+                      OMX_MAX_STRINGNAME_SIZE) ||
+      ! strncmp(cmp_name,
+                      "OMX.qti.video.decoder.vc1sw",
+                      OMX_MAX_STRINGNAME_SIZE) ) {
+
+     /**
+     * Only the following state transitions are allowed in omx swvdec
+     *
+     * LOADED -> IDLE -> EXECUTING
+     * LOADED <- IDLE <- EXECUTING
+     */
+      gst_omx_component_set_state (self->dec, OMX_StateIdle);
+    } else {
+      gst_omx_component_set_state (self->dec, OMX_StatePause);
+    }
     gst_omx_component_get_state (self->dec, GST_CLOCK_TIME_NONE);
   }
 #if defined (USE_OMX_TARGET_RPI) && defined (HAVE_GST_GL)
-- 
2.7.4

